src/platform/plugins/shared/controls/public/controls/data_controls/options_list_control/get_options_list_control_factory.tsx

```diff
@@ -349,111 +333,25 @@ export const getOptionsListControlFactory = (): DataControlFactory<
-            if (!existsSelected) {
-              selectionsManager.api.setSelectedOptions([]);
-              temporaryStateManager.api.setInvalidSelections(new Set([]));
-            }
-          } else if (showOnlySelected || selectedOptions.includes(keyAsType)) {
-            componentApi.deselectOption(key);
-          } else if (singleSelect) {
-            // replace selection
-            selectionsManager.api.setSelectedOptions([keyAsType]);
-            if (existsSelected) selectionsManager.api.setExistsSelected(false);
-          } else {
-            // select option
-            if (existsSelected) selectionsManager.api.setExistsSelected(false);
-            selectionsManager.api.setSelectedOptions(
-              selectedOptions ? [...selectedOptions, keyAsType] : [keyAsType]
-            );
-          }
-        },
-        sort$,
-        setSort: (sort: OptionsListSortingType | undefined) => {
-          sort$.next(sort);
-        },
-        selectAll: (keys: string[]) => {
-          const field = api.field$.getValue();
-          if (keys.length < 1 || !field) {
-            api.setBlockingError(
-              new Error(OptionsListStrings.control.getInvalidSelectionMessage())
-            );
-            return;
-          }
-
-          const selectedOptions = selectionsManager.api.selectedOptions$.getValue() ?? [];
-          const newSelections = keys.filter((key) => !selectedOptions.includes(key as string));
-          selectionsManager.api.setSelectedOptions([...selectedOptions, ...newSelections]);
-        },
-        deselectAll: (keys: string[]) => {
-          const field = api.field$.getValue();
-          if (keys.length < 1 || !field) {
-            api.setBlockingError(
-              new Error(OptionsListStrings.control.getInvalidSelectionMessage())
-            );
-            return;
-          }
-
-          const selectedOptions = selectionsManager.api.selectedOptions$.getValue() ?? [];
-          const remainingSelections = selectedOptions.filter(
-            (option) => !keys.includes(option as string)
-          );
-          selectionsManager.api.setSelectedOptions(remainingSelections);
-        },
+        deselectOption: (key) =>
+          deselectOption({ api, selectionsManager, temporaryStateManager, key }),
+        makeSelection: (key: string | undefined, showOnlySelected: boolean) =>
+          makeSelection({
+            api,
+            selectionsManager,
+            temporaryStateManager,
+            editorStateManager,
+            key,
+            showOnlySelected,
+          }),
+        selectAll: (keys: string[]) => selectAll({ api, keys, selectionsManager }),
+        deselectAll: (keys: string[]) => deselectAll({ api, keys, selectionsManager }),
+        allowExpensiveQueries$,
       };
 
-      if (selectionsManager.api.hasInitialSelections) {
-        await dataControlManager.api.untilFiltersReady();
-      }
```

Comment:
With `isFetchPaused$` on the Dashboard side, we no longer need to await the filters being ready.